advisor/src/main/kotlin/advisors/Osv.kt

```diff
@@ -0,0 +1,199 @@
+import org.ossreviewtoolkit.model.AdvisorCapability
+import org.ossreviewtoolkit.model.AdvisorDetails
+import org.ossreviewtoolkit.model.AdvisorResult
+import org.ossreviewtoolkit.model.AdvisorSummary
+import org.ossreviewtoolkit.model.Identifier
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.VulnerabilityReference
+import org.ossreviewtoolkit.model.config.AdvisorConfiguration
+import org.ossreviewtoolkit.utils.common.collectMessages
+import org.ossreviewtoolkit.utils.common.enumSetOf
+import org.ossreviewtoolkit.utils.common.withoutPrefix
+import org.ossreviewtoolkit.utils.ort.OkHttpClientHelper
+import org.ossreviewtoolkit.utils.ort.log
+
+/**
+ * An advice provider that obtains vulnerability from OSV.dev https://osv.dev/.
+ */
+class Osv(name: String, advisorConfiguration: AdvisorConfiguration) : AdviceProvider(name) {
+    class Factory : AbstractAdviceProviderFactory<Osv>("Osv") {
+        override fun create(config: AdvisorConfiguration): Osv =
+            Osv(providerName, config)
+    }
+
+    override val details: AdvisorDetails = AdvisorDetails(providerName, enumSetOf(AdvisorCapability.VULNERABILITIES))
+
+    private val service = OsvService(
+        serverUrl = advisorConfiguration.osv?.serverUrl ?: OsvApiClient.SERVER_URL_PRODUCTION,
+        httpClient = OkHttpClientHelper.buildClient()
+    )
+
+    override suspend fun retrievePackageFindings(packages: List<Package>): Map<Package, List<AdvisorResult>> {
+        val startTime = Instant.now()
+
+        val vulnerabilitiesForPackage = getVulnerabilitiesForPackage(packages)
+
+        log.info { "Found ${vulnerabilitiesForPackage.values.flatten().distinct().size} distinct vulnerabilities." }
+
+        val affectedPackages = vulnerabilitiesForPackage.filter { it.value.isNotEmpty() }.keys
+        if (affectedPackages.isNotEmpty()) {
+            log.info {
+                "Affected packages:\n\n${affectedPackages.joinToString("\n") { it.toCoordinates() }}\n"
+            }
+        }
+
+        return packages.associateWith { pkg ->
+            val vulnerabilities = vulnerabilitiesForPackage[pkg.id].orEmpty().map { it.toOrtVulnerability() }
+
+            if (vulnerabilities.isNotEmpty()) {
+                listOf(
+                    AdvisorResult(
+                        advisor = details,
+                        summary = AdvisorSummary(
+                            startTime = startTime,
+                            endTime = Instant.now()
+                        ),
+                        vulnerabilities = vulnerabilities
+                    )
+                )
+            } else {
+                emptyList()
+            }
+        }
+    }
+
+    private fun getVulnerabilitiesForPackage(packages: List<Package>): Map<Identifier, List<Vulnerability>> {
+        val vulnerabilityIdsForPackageId = getVulnerabilityIdsForPackages(packages)
+        val allVulnerabilityIds = vulnerabilityIdsForPackageId.values.flatten().toSet()
```

Comment:
Nit: Use `flattenTo(mutableSet())`.

Reply:
I cannot find that function, can you give a hint please?

Reply:
Sorry, it would be `flatMapTo(mutableSetOf()) { it }`, but that's not nicer than the current code. So let's keep the current code even if it might be a tad less performant.